File: js/scrubbed.ts
// agent.ts

import 'dotenv/config'
import { ChatOpenAI } from "@langchain/openai";
import { MemorySaver, StateGraph, START, END, type StreamMode,
    Annotation, type Messages, messagesStateReducer } 
    from "@langchain/langgraph";
import { BaseMessage, HumanMessage, 
  AIMessage, AIMessageChunk } from "@langchain/core/messages";
import { LangChainTracer } from "@langchain/core/tracers/tracer_langchain";
import * as readline from 'readline';
import { Client } from "langsmith";

// Initialize LangSmith client
const langsmithClient = new Client({
  hideInputs: (inputs) => ({}),
  hideOutputs: (outputs) => ({}),
});
const tracer = new LangChainTracer({
  client: langsmithClient,
});

// Create readline interface
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Create a promise-based prompt function
const prompt = (question: string): Promise<string> => {
    return new Promise((resolve) => {
        rl.question(question, (answer) => {
            resolve(answer);
        });
    });
};

// Define the state schema (messages array)
const StateAnnotation = Annotation.Root({
    messages: Annotation<BaseMessage[], Messages>({
      reducer: messagesStateReducer,
    }),
});

const llm = new ChatOpenAI({ model: "gpt-4o-mini" });
const checkpointer = new MemorySaver();

async function chatbot(state: any) {
    const response = await llm.invoke(state.messages);
    return { messages: [...state.messages, response] };
}

const graphBuilder = new StateGraph(StateAnnotation);
graphBuilder
.addNode("chatbot", chatbot)
.addEdge(START, "chatbot")
.addEdge("chatbot", END);

const graph = graphBuilder.compile({checkpointer}).withConfig({
  runName: "scrubbed",
  callbacks: [tracer],
});

async function runChat() {
    const config = {
      configurable: {
        thread_id: "your_thread_id",
      },
    };
    try {
        while (true) {
            const userInput = await prompt("User: ");
            if (!userInput) {
                console.log("Goodbye!");
                break;
            }
            if (["quit", "exit", "q"].includes(userInput.toLowerCase())) {
                console.log("Goodbye!");
                break;
            }
            const inputs = {messages: [new HumanMessage(userInput)]}
            for await (const chunk of await graph.stream(inputs, { ...config, streamMode: ["events", "messages"] as StreamMode[] })) {
              if (chunk[0] instanceof AIMessageChunk) {
                process.stdout.write(String(chunk[0].content));
              } else if (chunk[0] === 'messages' && chunk[1][0] instanceof AIMessageChunk) {
                process.stdout.write(String(chunk[1][0].content));
              }
            }
            console.log(); // Add newline after response
        }
    } finally {
        rl.close();
    }
}
  
runChat();

File: rats/scrubbed.py
import random 
import asyncio
from dotenv import load_dotenv
from typing import List
from typing_extensions import TypedDict
from contextlib import asynccontextmanager

from langchain.schema import Document
from langchain_openai import ChatOpenAI
from langchain_core.messages import HumanMessage, AIMessage, ToolMessage
from langchain_community.tools.tavily_search import TavilySearchResults

from langgraph.graph import StateGraph, START, END
from langgraph.types import Command, Interrupt, interrupt
from langgraph.checkpoint.sqlite.aio import AsyncSqliteSaver

from langsmith import Client
from langsmith.run_helpers import tracing_context

from IPython.display import Image, display


# Load environment variables
load_dotenv(dotenv_path=".env", override=True)

# Initialize LLM
llm = ChatOpenAI(model_name="gpt-4o", temperature=0)

# Initialize web search tool
web_search_tool = TavilySearchResults(max_results=3)

# Configure LangSmith client to hide data
langsmith_client = Client(
    hide_inputs=lambda inputs: {},
    hide_outputs=lambda outputs: {}
)

class GraphState(TypedDict):
    """
    Represents the state of our graph.

    """
    question: str
    query: str
    documents: List[str]
    messages: List[str]


class InputState(TypedDict):
    question: str


async def clarify(state: GraphState):
    question = state["question"]
    prompt = """You are an expert in Google Search. You know the optimal query for any question.
    Your job is to find the perfect search query for the question.

    Question: {question} 
    Answer:"""
    formatted = prompt.format(question=question)
    generation = await llm.ainvoke([HumanMessage(content=formatted)])
    return {"query": generation.content, "question": question}


async def search(state: GraphState):
    query = state["query"]
    documents = state.get("documents", [])

    web_docs = await web_search_tool.ainvoke({"query": query})
    for d in web_docs:
        web_results = Document(page_content=d["content"])
        documents.append(web_results)

    return {"documents": documents}



async def answer(state: GraphState):
    question = state["question"]
    documents = state.get("documents", [])
   
    prompt = """You are a professor and expert in explaining complex topics in a way that is easy to understand. 
    Your job is to summarize the answer to a question based on relevant background context. 

    Question: {question} 

    Context: {context}

    Answer:"""
    formatted = prompt.format(question=question, context="\n".join([d.page_content for d in documents]))
    generation = await llm.ainvoke([HumanMessage(content=formatted)])
    return {"messages": [generation]}

@asynccontextmanager
async def make_graph(memory):
    graph = StateGraph(GraphState, input=InputState)

    graph.add_node("clarify", clarify)
    graph.add_node("search", search, metadata={"category": "docs"})
    graph.add_node("answer", answer)

    graph.add_edge(START, "clarify")
    graph.add_edge("clarify", "search")
    graph.add_edge("search", "answer")
    graph.add_edge("answer", END)
    with tracing_context(client=langsmith_client):
        yield graph.compile(checkpointer=memory)


def print_messages(response):
    if isinstance(response, dict):
        for key in response:
            if key == "documents":
                print("# of Documents: " + str(len(response["documents"])))
            elif key == "messages":
                print("messages: " + response["messages"][-1].content)
            else:
                print(key + ": " + str(response[key]))


async def run(graph: StateGraph):
    state: GraphState = {
        "messages": [],
    }

    thread_id = random.randint(0, 1000000)
    config = {
        "configurable": {
            "thread_id": str(thread_id),
            "checkpoint_ns": "para",
        },
        "langsmith_client": langsmith_client
    }
    while True:
        user = input('User (q to quit): ')
        if user in {'q', 'Q'}:
            print('AI: Goodbye!')
            break
        state["question"] = user
        turn_input = state

        try:
            async for chunk in graph.astream(turn_input, config, stream_mode=["messages"]):
                if isinstance(chunk, tuple) and isinstance(chunk[0], AIMessage):
                    print(chunk[0].content, flush=True, end="")
            print("\n")
        except Exception as e:
            print(f"Error: {str(e)}")
            raise e


async def main():            
    async with AsyncSqliteSaver.from_conn_string(":memory:") as memory:
        async with make_graph(memory) as graph:
            await run(graph)

if __name__ == "__main__":
    asyncio.run(main())